src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp

```diff
@@ -14,89 +14,650 @@
 
 #pragma once
 
-#include <bsoncxx/config/prelude.hpp>
+#include <exception>
+#include <initializer_list>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+#include "./operators.hpp"
+#include "./type_traits.hpp"
 
-#if defined(BSONCXX_POLY_USE_MNMLSTC)
+// Try to detect std::optional. We'll reuse some components from there for compatibility
+// with C++17 users.
+#define BSONCXX_HAVE_STDLIB_OPTIONAL 0
 
-#include <core/optional.hpp>
+#ifdef __has_include
+#if __has_include(<version>)
+#include <version>
+#if defined(__cpp_lib_optional) && __cpp_lib_optional >= 201606L
+#undef BSONCXX_HAVE_STDLIB_OPTIONAL
+#define BSONCXX_HAVE_STDLIB_OPTIONAL 1
+#include <optional>
+#endif
+#endif
+#endif
+
+#include <bsoncxx/config/prelude.hpp>
 
 namespace bsoncxx {
+
 namespace v_noabi {
+
 namespace stdx {
 
-using ::core::make_optional;
-using ::core::nullopt;
-using ::core::nullopt_t;
-using ::core::optional;
+/**
+ * @brief Implementation of an std::optional-like class template
+ *
+ * Presents mostly the same interface as std::optional from C++17.
+ *
+ * @tparam T The type being made "optional"
+ */
+template <typename T>
+class optional;
 
-}  // namespace stdx
-}  // namespace v_noabi
-}  // namespace bsoncxx
+#if BSONCXX_HAVE_STDLIB_OPTIONAL
+// If we have access to the stdlib component, re-use those so that users can use
+// the same types and constants defined in <optional>
+using std::bad_optional_access;
+using std::in_place;
+using std::in_place_t;
+using std::nullopt;
+using std::nullopt_t;
```

Comment:
This seems to conflict with the intended purpose of the polyfill. If these C++17 features are available, the polyfill is already rendered unnecessary (straight up use `std::optional<T>` instead).

Recommend removing the partial-use of stdlib `<optional>` features and instead bundle the `using std::*;` into CXX-2796 work (pre-C++17: all our-impl; post-C++17: all stdlib).